File: src/3/schedule.js
const express = require('express');
const multer = require('multer');
const fs = require('fs');
const csv = require('csv-parser');
const path = require('path');
const app = express();
const port = 3000;

// Serve static files (for our HTML form)
app.use(express.static('public'));

// Configure multer for file uploads (files stored temporarily in 'uploads' folder)
const upload = multer({ dest: 'uploads/' });

// Global definitions for days and shifts
const DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"];
const SHIFTS = ["morning", "afternoon", "evening"];

/**
 * Extract the primary shift keyword from a human‑friendly string.
 * E.g., "Morning (8:00 AM - 12:00 PM)" returns "morning".
 */
function extractShift(cellValue) {
  const value = cellValue.trim().toLowerCase();
  if (value.includes("morning")) return "morning";
  if (value.includes("afternoon")) return "afternoon";
  if (value.includes("evening")) return "evening";
  return "";
}

/**
 * Parse the CSV file and build an employee data object.
 * Expected CSV columns include ranked preferences for each day,
 * e.g., Monday_1, Monday_2, Monday_3, etc.
 */
function parseCSV(filePath) {
  return new Promise((resolve, reject) => {
    const employees = {};
    fs.createReadStream(filePath)
      .pipe(csv())
      .on('data', (row) => {
        const name = row['name'];
        const preferences = {};
        for (const day of DAYS) {
          const prefList = [];
          // Expecting columns: day_1, day_2, day_3
          for (const rank of ['_1', '_2', '_3']) {
            const key = day + rank;
            if (row[key]) {
              const shift = extractShift(row[key]);
              if (shift) prefList.push(shift);
            }
          }
          preferences[day] = prefList;
        }
        employees[name] = { preferences: preferences, days_worked: 0 };
      })
      .on('end', () => resolve(employees))
      .on('error', (err) => reject(err));
  });
}

/**
 * Generate the schedule using bonus logic with ranked preferences.
 */
function generateSchedule(employees) {
  // Initialize the schedule: for each day, each shift is an empty array.
  const schedule = {};
  for (const day of DAYS) {
    schedule[day] = {};
    for (const shift of SHIFTS) {
      schedule[day][shift] = [];
    }
  }

  // Process each day
  for (const day of DAYS) {
    const unassigned = [];
    // First pass: try to assign employees based on their ranked preferences
    for (const emp in employees) {
      let assigned = false;
      for (const pref of employees[emp].preferences[day] || []) {
        if (schedule[day][pref].length < 2) {
          schedule[day][pref].push(emp);
          employees[emp].days_worked++;
          assigned = true;
          break;
        }
      }
      if (!assigned) {
        unassigned.push(emp);
      }
    }

    // Second pass: assign unassigned employees to any available shift on the same day
    for (const emp of unassigned) {
      let assigned = false;
      for (const shift of SHIFTS) {
        if (schedule[day][shift].length < 2) {
          schedule[day][shift].push(emp);
          employees[emp].days_worked++;
          assigned = true;
          break;
        }
      }
      // If still not assigned, try the next day (if possible)
      if (!assigned) {
        const dayIndex = DAYS.indexOf(day);
        if (dayIndex < DAYS.length - 1) {
          const nextDay = DAYS[dayIndex + 1];
          for (const shift of SHIFTS) {
            if (schedule[nextDay][shift].length < 2) {
              schedule[nextDay][shift].push(emp);
              employees[emp].days_worked++;
              break;
            }
          }
        }
      }
    }

    // Finally, fill any remaining vacancies on the day by randomly selecting eligible employees.
    for (const shift of SHIFTS) {
      while (schedule[day][shift].length < 2) {
        // Find employees not already scheduled on this day.
        const assignedToday = [].concat(...SHIFTS.map(s => schedule[day][s]));
        const available = Object.keys(employees).filter(emp => !assignedToday.includes(emp));
        if (available.length > 0) {
          const chosen = available[Math.floor(Math.random() * available.length)];
          schedule[day][shift].push(chosen);
          employees[chosen].days_worked++;
        } else {
          break;
        }
      }
    }
  }
  return schedule;
}

/**
 * Convert the schedule object into an HTML table.
 */
function scheduleToHTML(schedule) {
  let html = '<h2>Final Weekly Schedule</h2>';
  html += '<table border="1" cellspacing="0" cellpadding="5">';
  html += '<tr><th>Day</th><th>Morning</th><th>Afternoon</th><th>Evening</th></tr>';
  for (const day of DAYS) {
    html += `<tr><td>${day}</td>`;
    for (const shift of SHIFTS) {
      html += `<td>${schedule[day][shift].join(", ")}</td>`;
    }
    html += '</tr>';
  }
  html += '</table>';
  return html;
}

// Route: Serve the main HTML page.
app.get('/', (req, res) => {
  res.sendFile(path.join(__dirname, 'public', 'index.html'));
});

// Route: Handle CSV file upload, process scheduling, and display results.
app.post('/upload', upload.single('csvfile'), async (req, res) => {
  try {
    const filePath = req.file.path;
    const employees = await parseCSV(filePath);
    const schedule = generateSchedule(employees);
    const htmlTable = scheduleToHTML(schedule);
    // Delete the uploaded file.
    fs.unlinkSync(filePath);
    res.send(`
      <html>
        <head><title>Final Schedule</title></head>
        <body>
          ${htmlTable}
          <br>
          <a href="/">Go Back</a>
        </body>
      </html>
    `);
  } catch (err) {
    res.status(500).send("Error processing file: " + err);
  }
});

// Start the server.
app.listen(port, () => {
  console.log(`Server running at http://localhost:${port}`);
});

File: src/3/schedule.py
import tkinter as tk
from tkinter import ttk, filedialog, messagebox
import csv
import random

# Global definitions for days and shifts
DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
SHIFTS = ["morning", "afternoon", "evening"]

def extract_shift(cell_value):
    """
    Extract the primary shift keyword from a human-friendly string.
    For example, "Morning (8:00 AM - 12:00 PM)" returns "morning".
    """
    value = cell_value.strip().lower()
    if "morning" in value:
        return "morning"
    elif "afternoon" in value:
        return "afternoon"
    elif "evening" in value:
        return "evening"
    else:
        return ""

class ScheduleApp:
    def __init__(self, master):
        self.master = master
        master.title("Employee Scheduling App")
        master.geometry("400x150")
        
        self.csv_path = None
        self.employees = {}  # will hold employee data from CSV
        self.schedule = {}   # final schedule data
        
        # Create and pack UI components
        self.label = tk.Label(master, text="Load your employee shift CSV (with ranked preferences)")
        self.label.pack(pady=5)
        
        self.load_button = tk.Button(master, text="Load CSV", command=self.load_csv)
        self.load_button.pack(pady=5)
        
        self.generate_button = tk.Button(master, text="Generate Schedule", command=self.generate_schedule, state=tk.DISABLED)
        self.generate_button.pack(pady=5)
    
    def load_csv(self):
        # Open file dialog to select CSV file
        path = filedialog.askopenfilename(title="Select CSV File", filetypes=(("CSV Files", "*.csv"),))
        if path:
            self.csv_path = path
            messagebox.showinfo("File Loaded", f"Loaded file:\n{path}")
            self.generate_button.config(state=tk.NORMAL)
    
    def generate_schedule(self):
        if not self.csv_path:
            messagebox.showerror("Error", "No CSV file loaded.")
            return
        
        # Read employee data from CSV file (expecting bonus columns, e.g., Monday_1, Monday_2, Monday_3, etc.)
        self.employees = {}
        try:
            with open(self.csv_path, 'r', newline='') as csvfile:
                reader = csv.DictReader(csvfile)
                for row in reader:
                    name = row['name']
                    # For each day, create a list of ranked preferences (first, second, third)
                    preferences = {}
                    for day in DAYS:
                        pref_list = []
                        for rank in ['_1', '_2', '_3']:
                            key = day + rank
                            if key in row:
                                shift = extract_shift(row[key])
                                if shift:
                                    pref_list.append(shift)
                        preferences[day] = pref_list
                    self.employees[name] = {"preferences": preferences, "days_worked": 0}
        except Exception as e:
            messagebox.showerror("Error", f"Failed to read CSV file:\n{e}")
            return
        
        # Initialize the schedule dictionary for each day and shift
        self.schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS}
        
        # Process scheduling for each day
        for day in DAYS:
            unassigned = []  # Employees who could not get any of their ranked preferences
            
            # First pass: assign employees based on their ranked preferences (in order)
            for emp, details in self.employees.items():
                assigned = False
                for pref in details["preferences"][day]:
                    if len(self.schedule[day][pref]) < 2:
                        self.schedule[day][pref].append(emp)
                        self.employees[emp]["days_worked"] += 1
                        assigned = True
                        break
                if not assigned:
                    unassigned.append(emp)
            
            # Second pass: assign unassigned employees to any available shift on the same day
            for emp in unassigned:
                assigned = False
                for shift in SHIFTS:
                    if len(self.schedule[day][shift]) < 2:
                        self.schedule[day][shift].append(emp)
                        self.employees[emp]["days_worked"] += 1
                        assigned = True
                        break
                # If still not assigned, try assigning on the next day (if possible)
                if not assigned:
                    day_index = DAYS.index(day)
                    if day_index < len(DAYS) - 1:
                        next_day = DAYS[day_index + 1]
                        for shift in SHIFTS:
                            if len(self.schedule[next_day][shift]) < 2:
                                self.schedule[next_day][shift].append(emp)
                                self.employees[emp]["days_worked"] += 1
                                break
            
            # Finally, fill any remaining vacancies on the day by randomly selecting eligible employees
            for shift in SHIFTS:
                while len(self.schedule[day][shift]) < 2:
                    # Find employees not already scheduled on this day
                    assigned_today = []
                    for s in SHIFTS:
                        assigned_today.extend(self.schedule[day][s])
                    available = [emp for emp in self.employees if emp not in assigned_today]
                    if available:
                        chosen = random.choice(available)
                        self.schedule[day][shift].append(chosen)
                        self.employees[chosen]["days_worked"] += 1
                    else:
                        break
        
        # Once scheduling is complete, display the schedule
        self.display_schedule()
    
    def display_schedule(self):
        # Create a new window for displaying the final schedule table
        win = tk.Toplevel(self.master)
        win.title("Final Weekly Schedule")
        win.geometry("600x300")
        
        # Create a Treeview widget for the table
        tree = ttk.Treeview(win)
        tree["columns"] = ("Morning", "Afternoon", "Evening")
        tree.heading("#0", text="Day")
        tree.column("#0", width=100)
        for col in tree["columns"]:
            tree.heading(col, text=col)
            tree.column(col, width=150)
        
        # Insert schedule data into the Treeview
        for day in DAYS:
            morning = ", ".join(self.schedule[day]["morning"])
            afternoon = ", ".join(self.schedule[day]["afternoon"])
            evening = ", ".join(self.schedule[day]["evening"])
            tree.insert("", "end", text=day, values=(morning, afternoon, evening))
        
        tree.pack(fill=tk.BOTH, expand=True, padx=10, pady=10)
        
        # Add a button to close the schedule window
        close_btn = tk.Button(win, text="Close", command=win.destroy)
        close_btn.pack(pady=5)

if __name__ == "__main__":
    root = tk.Tk()
    app = ScheduleApp(root)
    root.mainloop()
